### PR TITLE
Implement a proper GlyphCache interface

### PR DESCRIPTION
**Pull request by redstarcoder**
This is related to #121 and I think #126. This is untested. This allows `glyphCache` to be replaced and restored.

### CODE REVIEW

**Comment by gerald1248** [comment]
@redstarcoder thanks so much for this. I'll test with my application and the sync test & follow up here. (If we get the test to run with -race we'd have coverage for GlyphCache too.)

**Comment by gerald1248** [comment]
Not sure where I'm going wrong but I've tried to compile my test `sync_test.go` against the PR without success.

```
$ go test sync_test.go
# command-line-arguments_test
./sync_test.go:128: unknown draw2dbase.Glyph field 'path' in struct literal
FAIL	command-line-arguments [build failed]
```
I must be overlooking something dead obvious:

```
101 //testGlyphCache tracks @redstarcoder's glyph cache interface PR
102 type testGlyphCache struct {
103   glyphMap map[string]map[rune]*draw2dbase.Glyph
104 }
105
106 // Fetch fetches a glyph from the cache, storing with Render first if it doesn't already exist
107 func (cache *testGlyphCache) Fetch(gc draw2d.GraphicContext, fontName string, chr rune) *draw2dbase.Glyph {
108   var mu sync.Mutex
109   mu.Lock()
110   if cache.glyphMap[fontName] == nil {
111     cache.glyphMap[fontName] = make(map[rune]*draw2dbase.Glyph, 60)
112   }
113   if cache.glyphMap[fontName][chr] == nil {
114     cache.glyphMap[fontName][chr] = cache.Render(gc, fontName, chr)
115   }
116   mu.Unlock()
117   return cache.glyphMap[fontName][chr].Copy()
118 }
119
120 // Render renders a glyph then returns it
121 func (cache *testGlyphCache) Render(gc draw2d.GraphicContext, fontName string, chr rune) *draw2dbase.Glyph {
122   gc.Save()
123   defer gc.Restore()
124   gc.BeginPath()
125   width := gc.CreateStringPath(string(chr), 0, 0)
126   path := gc.GetPath()
127   return &draw2dbase.Glyph{
128     path:  &path,
129     Width: width,
130   }
131 }
132
133 var (
134   syncGlyphCache = &testGlyphCache{make(map[string]map[rune]*draw2dbase.Glyph)}
135   glyphCache draw2dbase.GlyphCache = syncGlyphCache
136 )
```

I wonder if we're going in the wrong direction. It seems really complex making the library be threadsafe. Should we perhaps avoid globals altogether, attach all caches to a given context and allow the library user to get/set caches to reuse them if they happen to need many non-concurrent contexts?

**Comment by llgcode** [comment]
Hi @gerald1248 
path is a private field so you can't access it from another package.

I agree we need to avoid globals. But in the same time we need to keep the api compatible. So we will do it in 2 steps (for font cache). First add mutex to avoid data race second make a new Api so that it allow to configure in each Gc the right cache.

@redstarcoder 
I propose to change GlyphCache Api now. We need to remove the global and add a setter on StackGraphicContext.

**Comment by redstarcoder** [comment]
@llgcode I agree, I'll make the changes!

**Comment by redstarcoder** [comment]
@llgcode I made the `GlyphCache` settable per `GraphicContext`. This required a bit more changes, but I think I made it comfortable. Currently by default new gcs will still use the global cache, but the cache is settable now so a user could maintain their own caches per GraphicContext, or use mutex locks. I think this system is much better now.

**Comment by redstarcoder** [comment]
@gerald1248 with the changes I mentioned above, if you set a new `GlyphCache` for each `GraphicContext`, you only need to mutex lock the fill/stroke routines.

@llgcode now that I'm thinking of multiple `GlyphCache`s ... perhaps a `NewGlyphCache` function is appropriate inside `draw2dbase/text.go`? That way people won't need their own `Fetch` and `Render` methods (which could very well be exact copies).

File: gc.go [hunk 1]
```diff
@@ -52,6 +52,10 @@ type GraphicContext interface {
 	GetFontData() FontData
 	// GetFontName gets the current FontData as a string
 	GetFontName() string
+	// SetGlyphCache changes the glyph cache backend used by the GraphicContext
+	SetGlyphCache(cache GlyphCache)
```
Comment (llgcode) [inline gc.go]: I think this function don't have to be in the Gc interface. It is useful only in some implementation of Gc not all.

**Comment by llgcode** [comment]
@redstarcoder yes a public constructor NewGlyphCache will be useful.

**Comment by gerald1248** [comment]
@redstarcoder ah I was being optimistic and assuming concurrent reading/copying is allowed. Thanks to both of you for the sterling work on the library!

**Comment by redstarcoder** [comment]
@gerald1248 you can concurrently read, however if you keep the read outside of the mutex lock, you have no guarentee anymore that the read doesn't happen while a write occurs :p. Always happy to help out in useful projects!

@llgcode I've thought about it and I agree, the cache can always be set before the gc becomes an interface. I'll begin work!

**Comment by redstarcoder** [comment]
@llgcode finished! This also removes `draw2dbase.FetchGlyph` which while public, is probably pretty safe to remove.

**Comment by llgcode** [comment]
thanks @redstarcoder.
Great just for Chrismas. Just one thing, The GlyphCache could be attached directly to gc, I'm not sure it's necessary to set it on the current context.
cheers

**Comment by redstarcoder** [comment]
@llgcode yeah! :) it really snuck on me this year hehe.

That makes sense. So move `SetGlyphCache` directly into `draw2dimg/ftgc.go` and `draw2dgl/gc.go`? I'll do that! That way save/restore calls won't change caches unexpectedly.

Have a merry Christmas! I might be a little less available for the next day or two, but I'll be back to programming ASAP!

**Comment by gerald1248** [comment]
@redstarcoder thanks for the code sample. It all seems very obvious and my original version very obviously wrong now :) Thanks!

**Comment by llgcode** [comment]
Merry chrismas :-)

**Comment by llgcode** [comment]
thanks @redstarcoder  I've merged it using a local branch.

**Pull request closed without merging**